R/MsBackendTimsTof-functions.R

```diff
@@ -357,3 +375,53 @@ MsBackendTimsTof <- function() {
         as(res, "DataFrame")
     else DataFrame()
 }
+
+#' @importFrom BiocParallel bpmapply bpparam
+.calculate_core_ms2_information <- function(x){
+    if (is.null(names(x@fileNames))){
+        return(matrix(numeric(), nrow = 0, ncol = length(.TIMSTOF_MS2_COLUMNS)))
+    }
+    tbls <- bpmapply(FUN = .do_calculate_core_ms2_information,
+                     x = names(x@fileNames),
+                     indices = lapply(split(x@indices[,1:2], f = x@indices[,3]),
+                                      matrix, ncol = 2),
+                     SIMPLIFY = FALSE,
+                     BPPARAM = bpparam())
+    output <- do.call(rbind, tbls)
+    output[order(order(x@indices[,3])), ]
+}
+
+#' @importFrom opentimsr OpenTIMS CloseTIMS table2df
+#' @importFrom MsCoreUtils between
+.do_calculate_core_ms2_information <- function(x, indices){
+    tms <- opentimsr::OpenTIMS(x)
+    on.exit(opentimsr::CloseTIMS(tms))
+    if (missing(indices)) {
+        indices <- unique(.query_tims(tms, frames = tms@frames$Id,
+                                      columns = c("frame", "scan")))
+    }
+    output <- matrix(NA_real_,
+                     nrow = nrow(indices),
+                     ncol = length(.TIMSTOF_MS2_COLUMNS))
+    tbl <- opentimsr::table2df(tms, c("PASEFFrameMsMsInfo", "Precursors"))
+    for (i in seq(nrow(tbl[[1]]))) {
+        row <- tbl[[1]][i, ]
+        prec <- tbl[[2]][row$Precursor,]
+        target_rows <- which(indices[,1] == row$Frame &
+                             MsCoreUtils::between(indices[,2],
+                                                  c(row$ScanNumBegin,
+                                                    row$ScanNumEnd)))
+        if (length(target_rows)) {
+            output[target_rows, ] <- 
+                rep(c(prec$MonoisotopicMz,                       ## precursorMz
+                         prec$Charge,                            ## precursorCharge
+                         prec$Intensity,                         ## precursorIntensity
+                         row$CollisionEnergy,                    ## collisionEnergy
+                         row$IsolationMz - row$IsolationWidth,   ## isolationWindowLowerMz
```

Comment:
Hm, I'm wondering how the isolation width is defined - just to be sure that it should not be isolationMz - isolationWidth/2 ... can you please check?